cpp/node_similarity_module/algorithms/node_similarity.hpp

```diff
@@ -0,0 +1,206 @@
+    if (denonominator == 0) {
+        return 0.0;
+    } else {
+        return elem_intersection.size() / (double) denonominator;
+    }
+}
+
+
+
+/*
+Calculates cosine similarity function between two nodes for a given property.
+*/
+double CosineFunc(const mgp::Node &node1, const mgp::Node &node2, const std::string &property) {
+    const auto &prop1_it = node1.GetProperty(property);
+    const auto &prop2_it = node2.GetProperty(property);
+    if (prop1_it.IsNull() || prop2_it.IsNull()) {
+        throw mgp::ValueException("All nodes should have property " + property);
+    }
+    const auto &prop1 = prop1_it.ValueList();
+    const auto &prop2 = prop2_it.ValueList();
+    double similarity = 0.0, node1_sum = 0.0, node2_sum = 0.0;
+    int size1 = prop1.Size(), size2 = prop2.Size();
+    if (size1 != size2) {
+        throw mgp::ValueException("Vectors should be of the same size.");
+    }
+    for (int i = 0; i < size1; ++i) {
+        double val1 = prop1[i].ValueDouble(), val2 = prop2[i].ValueDouble();
+        similarity += val1 * val2;
+        node1_sum += val1 * val1;
+        node2_sum += val2 * val2;
+    }
+    double denominator = sqrt(node1_sum) * sqrt(node2_sum);
+    if (denominator < 1e-9) {
+        return 0.0;
+    } else {
+        return similarity / denominator;
+    }
+}
+
+
+
+/*
+Extract node neighbours.
+*/
+std::set<uint64_t> GetNeighbors(std::unordered_map<uint64_t, std::set<uint64_t>> &neighbors, const mgp::Node &node) {
+    uint64_t node_id = node.Id().AsUint();
+    const auto &result_it = neighbors.find(node_id);
+    if (result_it == neighbors.end()) {
+        std::set<uint64_t> ns;
+        const auto &rels = node.OutRelationships();
+        for (const auto &rel: rels) {
+            ns.insert(rel.To().Id().AsUint());
+        }
+        neighbors[node_id] = ns;
+        return ns;
+    } else {
+        return result_it->second;
+    }
+}
+
+
+/*
+Calculates similiraty between pairs of nodes given by src_nodes and dst_nodes.
+*/
+std::vector<std::tuple<mgp::Node, mgp::Node, double>> CalculateSimilarityPairwise(const mgp::List &src_nodes, const mgp::List &dst_nodes, node_similarity_util::Similarity similarity_mode, std::string property = "") {
+    if (src_nodes.Size() != dst_nodes.Size()) {
+        throw mgp::ValueException("Arguments are of different size.");
+    }
+    int num_nodes = src_nodes.Size();
+    std::vector<std::tuple<mgp::Node, mgp::Node, double>> results;
+    std::unordered_map<uint64_t, std::set<uint64_t>> neighbors;
+    for (int i = 0; i < num_nodes; ++i) {
+        const mgp::Node &src_node = src_nodes[i].ValueNode(), &dst_node = dst_nodes[i].ValueNode();
+        double similarity = 0.0;
+        if (similarity_mode == node_similarity_util::Similarity::cosine) {
+            similarity = node_similarity_algs::CosineFunc(src_node, dst_node, property);
+        } else {
+            const auto &ns1 = GetNeighbors(neighbors, src_node);
```

Comment:
I think this can be simplified:

if (similarity_mode == node_similarity_util::Similarity::cosine) {
            similarity = node_similarity_algs::CosineFunc(src_node, dst_node, property);
            continue;
}
And here then do the rest of the logic without `else`. I was trying to think of the way to group everything together but couldn't do it in elegant way.

Reply:
nice catch!